[[ Bug 22032 ]] Use simple mapping in toUpper and toLower

montegoulding opened the pull request.

runrevmark commented on libfoundation/src/foundation-unicode.cpp:
```diff
@@ -771,28 +771,86 @@ bool MCUnicodeIsNormalisedNKD(const unichar_t *p_string, uindex_t p_length)
+            icu::Locale t_locale = nullptr;
+            if (p_locale != nullptr)
+            {
+                t_locale = MCLocaleGetICULocale(p_locale);
```
Does this cause a memory leak in icu::Locale objects?

montegoulding commented on libfoundation/src/foundation-unicode.cpp:
```diff
@@ -771,80 +771,142 @@ bool MCUnicodeIsNormalisedNKD(const unichar_t *p_string, uindex_t p_length)
+            icu::Locale t_locale = nullptr;
+            if (p_locale != nullptr)
+            {
+                t_locale = MCLocaleGetICULocale(p_locale);
```
@runrevmark maybe you deleted your comment but I got the email anyway? Anyway no this does not leak it's just an abstraction to access the icu locale in the `__MCLocale` struct.

runrevmark replied:
Indeed - that's why I deleted the comment - I missed the fact the locale-ref was passed in first time around :)

montegoulding commented:
I just noticed the milestone change so I'll rebase

montegoulding committed:
[[ Bug 22032 ]] Implement __MCUnicodeMapString

This patch implements a utility function for mapping a unicode string from one
form to another for a particular unicode character property.

montegoulding committed:
[[ Bug 22032 ]] Use __MCUnicodeMapString

This patch changes MCUnicodeLowercase, MCUnicodeUppercase and MCUnicodeTitlecase to
use the new __MCUnicodeMapString utility function. It also adds support for simple
mapping in those functions where the locale parameter is null.

montegoulding committed:
[[ Bug 22032 ]] Use simple mapping in toUpper and toLower functions

This patch changes the toUpper and toLower functions to use simple unicode
case mapping.
